# EXAMEN 2
## Cálculo aplicado a Machine Learning

**Nombre:** Juan José  
**Apellido 1:** Melo  
**Apellido 2:** Montenegro  
**Rama:** melo_montenegro

In [1]:
import numpy as np

## 1. Datos

Carga los datos de tiempos de pedido y separa el número de productos (`x`) del tiempo observado (`y`).

In [8]:
data = np.loadtxt(
    "../data/order_times.csv",
    delimiter=",",
    skiprows=1
)

x = data[:, 0]
y = data[:, 1]

print("x:", x)
print("y:", y)

x: [1. 2. 3. 4. 5. 6.]
y: [ 3.2  5.1  6.8  9.2 11.  13.1]


## 2. Función de predicción

Completa la función para calcular las predicciones del modelo lineal.

In [9]:
def predict(x, w, b):
    # Completar
    return w * x + b

## 3. Función de pérdida

Completa la función para calcular el error cuadrático medio (MSE).

In [10]:
def mse_loss(y_true, y_pred):
    # Completar
    return np.mean((y_true - y_pred) ** 2)

## 4. Derivada numérica

Implementa una aproximación numérica de la derivada usando la función `quadratic`.

In [11]:
def quadratic(x):
    return x**2


def numerical_derivative(
    f,
    x,
    h=1e-5
):
    # Completar
    return (f(x + h) - f(x)) / h

### 4.1. Prueba de derivada numérica

Usa `numerical_derivative(quadratic, 3)` para aproximar la derivada de $x^2$ en $x=3$.

In [12]:
derivative_at_3 = numerical_derivative(quadratic, 3)

print(
    "Derivada aproximada de x^2 en x=3:",
    derivative_at_3
)

Derivada aproximada de x^2 en x=3: 6.000009999951316


## 5. Gradientes de la regresión lineal

Completa la función para calcular los gradientes respecto a `w` y `b`.

In [13]:
def linear_regression_gradients(
    x,
    y,
    w,
    b
):
    # Completar

    n = len(x)
    y_pred = predict(x, w, b)

    dw = (-2/n) * np.sum(x * (y - y_pred))
    db = (-2/n) * np.sum(y - y_pred)

    return dw, db

## 6. Paso de Gradient Descent

Implementa un paso de actualización de los parámetros usando los gradientes y `learning_rate`.

In [14]:
def gradient_descent_step(
    x,
    y,
    w,
    b,
    learning_rate
):
    # Completar
    dw, db = linear_regression_gradients(x, y, w, b)
    w -= learning_rate * dw
    b -= learning_rate * db

    return w, b

## 7. Función de entrenamiento

Entrena la regresión lineal durante el número indicado de épocas.

In [15]:
def train_linear_regression(
    x,
    y,
    learning_rate=0.01,
    epochs=1000
):
    # TODO: entrenar el modelo y reemplazar estos valores

    w = 0.0
    b = 0.0

    for epoch in range(epochs):
        y_pred = predict(x, w, b)
        loss = mse_loss(y, y_pred)

        w, b = gradient_descent_step(x, y, w, b, learning_rate)

        if epoch % 100 == 0:
            print(epoch, loss, w, b)

    return w, b

## 8. Modelo inicial

Calcula la predicción y la pérdida iniciales usando los parámetros indicados.

In [16]:
initial_w = 0.0
initial_b = 0.0

# TODO: calcular la predicción inicial usando initial_w e initial_b
initial_prediction = predict(x, initial_w, initial_b)

# TODO: calcular la pérdida inicial
initial_loss = mse_loss(y, initial_prediction)

## 9. Entrenamiento del modelo

Entrena el modelo con los datos cargados.

In [17]:
w, b = train_linear_regression(
    x,
    y,
    learning_rate=0.01,
    epochs=1000
)

0 76.62333333333333 0.6806666666666666 0.16133333333333333
100 0.052853355326685524 2.086892743993415 0.6857335985893843
200 0.035039076633047984 2.0567840398866064 0.8146349374406334
300 0.026464706383949432 2.035895453097193 0.9040631238652027
400 0.022337689711930897 2.021403528884481 0.966105928063043
500 0.020351273670133625 2.011349432381723 1.0091495109269133
600 0.019395171719331566 2.0043741780950564 1.039011959114311
700 0.018934980640981067 1.9995349394473751 1.059729700352657
800 0.018713481429086586 1.9961776094165584 1.0741030966437293
900 0.018606869408171852 1.9938483866283359 1.0840749617502325


## 10. Modelo final

Calcula las predicciones y la pérdida después del entrenamiento, y muestra los resultados.

In [18]:
# TODO: calcular las predicciones finales
final_prediction = predict(x, w, b)

# TODO: calcular la pérdida final
final_loss = mse_loss(y, final_prediction)

print("Pérdida inicial:", initial_loss)
print("Pérdida final:", final_loss)
print("Valor final de w:", w)
print("Valor final de b:", b)

Pérdida inicial: 76.62333333333333
Pérdida final: 0.018555554900520545
Valor final de w: 1.9922458457901508
Valor final de b: 1.090935757182446


## 11. Nueva predicción

Predice el tiempo para un pedido con 7 productos usando los parámetros entrenados.

In [19]:
x_new = 7

prediction = predict(x_new, w, b)

print(
    "Predicción para un pedido con 7 productos:",
    prediction
)

Predicción para un pedido con 7 productos: 15.036656677713502


## 12. Preguntas de interpretación

1. ¿Qué representan `w` y `b` dentro de este problema?

   **Respuesta:** `w` es el cambio estimado en el tiempo por cada producto adicional; `b` es el tiempo estimado para un pedido con cero productos. En este entrenamiento, `w` es aproximadamente 1.99 y `b` aproximadamente 1.09.

2. ¿Qué representa la función de pérdida?

   **Respuesta:** Mide el error de las predicciones respecto a los valores reales.

3. ¿A qué valor debería aproximarse la derivada de $x^2$ en $x=3$?

   **Respuesta:** A 6, porque la derivada de $x^2$ es $2x$.

4. ¿Qué representa `dw`?

   **Respuesta:** Es la derivada de la pérdida respecto a `w`; indica cómo cambia la pérdida al modificar la pendiente del modelo.

5. ¿Qué representa `db`?

   **Respuesta:** Es la derivada de la pérdida respecto a `b`; indica cómo cambia la pérdida al modificar la intersección del modelo.

6. ¿Por qué Gradient Descent resta el gradiente en lugar de sumarlo?

   **Respuesta:** El gradiente apunta hacia el aumento más rápido de la pérdida; restarlo mueve los parámetros hacia una pérdida menor.

7. ¿Qué significa que la pérdida final sea menor que la pérdida inicial?

   **Respuesta:** Que las predicciones se acercan más a los valores reales.

8. ¿Qué efecto tiene el learning rate durante el entrenamiento?

   **Respuesta:** Determina el tamaño de cada actualización. Si es muy grande, puede sobrepasar el mínimo; si es muy pequeño, el aprendizaje será lento.